### PR TITLE
Support sort by key path

### PR DESCRIPTION
**Pull request by MarkRunWu**
connect #563

### CODE REVIEW

File: pkg/server/skydb/pq/builder/factory.go [hunk 1]
```diff
@@ -326,3 +332,34 @@ type joinedTable struct {
 func (a joinedTable) equal(b joinedTable) bool {
 	return a.secondaryTable == b.secondaryTable && a.primaryColumn == b.primaryColumn && a.secondaryColumn == b.secondaryColumn
 }
+
+func (f *predicateSqlizerFactory) NewSort(s skydb.Sort) (string, error) {
```
Comment (Steven-Chan) [inline pkg/server/skydb/pq/builder/factory.go]: This function is doing the same thing as `SortOrderBySQL`, so you either need this `NewSort` or `SortOrderBySQL`. It is so unpredictable now, you cannot tell which code path will be executed for each type of expression. Please combine them into one single function.

File: pkg/server/skydb/pq/builder/factory.go [hunk 2]
```diff
@@ -229,7 +230,7 @@ func (f *predicateSqlizerFactory) newExpressionSqlizer(expr skydb.Expression) (e
 		`unexpected expression type`)
 }
 
-func (f *predicateSqlizerFactory) newExpressionSqlizerForKeyPath(expr skydb.Expression) (expressionSqlizer, error) {
+func (f *predicateSqlizerFactory) newExpressionSqlizerForKeyPath(expr skydb.Expression, addJoinedFieldToExtraColumn bool) (expressionSqlizer, error) {
```
Comment (Steven-Chan) [inline pkg/server/skydb/pq/builder/factory.go]: See `Avoid Side Effects` in https://medium.com/javascript-scene/the-single-biggest-mistake-programmers-make-every-day-62366b432308.

In short, if a function returns something except error, it is expected not to have any side effects.

File: pkg/server/skydb/pq/builder/factory.go [hunk 3]
```diff
@@ -26,6 +26,7 @@ import (
 type PredicateSqlizerFactory interface {
 	UpdateTypemap(typemap skydb.RecordSchema) skydb.RecordSchema
 	AddJoinsToSelectBuilder(q sq.SelectBuilder) sq.SelectBuilder
+	NewSort(s skydb.Sort) (string, error)
```
Comment (Steven-Chan) [inline pkg/server/skydb/pq/builder/factory.go]: This is a `predicate` `sqlizer` factory, and this new function fits neither.

File: pkg/server/skydb/pq/builder/factory.go [hunk 4]
```diff
@@ -326,3 +332,34 @@ type joinedTable struct {
 func (a joinedTable) equal(b joinedTable) bool {
 	return a.secondaryTable == b.secondaryTable && a.primaryColumn == b.primaryColumn && a.secondaryColumn == b.secondaryColumn
 }
+
+func (f *predicateSqlizerFactory) NewSort(s skydb.Sort) (string, error) {
+	expr := s.Expression
+	if expr.IsKeyPath() && len(expr.KeyPathComponents()) == 2 {
```
Comment (Steven-Chan) [inline pkg/server/skydb/pq/builder/factory.go]: Please handle `len(expr.KeyPathComponents()) > 2` explicitly.

File: pkg/server/skydb/pq/builder/sort.go [hunk 5]
```diff
@@ -26,7 +26,11 @@ func SortOrderBySQL(alias string, sort skydb.Sort) (string, error) {
 
 	switch sort.Expression.Type {
 	case skydb.KeyPath:
-		expr = fullQuoteIdentifier(alias, sort.Expression.Value.(string))
+		if len(sort.Expression.KeyPathComponents()) == 2 {
```
Comment (Steven-Chan) [inline pkg/server/skydb/pq/builder/sort.go]: Please handle `len(expr.KeyPathComponents()) > 2` explicitly.

File: pkg/server/skydb/pq/recorddb.go [hunk 6]
```diff
@@ -599,8 +609,14 @@ func columnSqlizersForSelect(recordType string, typemap skydb.RecordSchema) map[
 				Value: column,
 			}
 		}
+		var sqlizer sq.Sqlizer
 
-		sqlizer := builder.NewExpressionSqlizer(recordType, fieldType, expr)
+		if fieldType.Type == skydb.TypeReference && len(expr.KeyPathComponents()) == 2 {
```
Comment (Steven-Chan) [inline pkg/server/skydb/pq/recorddb.go]: Please handle `len(expr.KeyPathComponents()) > 2` explicitly.

Comment (MarkRunWu) [reply pkg/server/skydb/pq/recorddb.go]: Updated

File: pkg/server/skydb/pq/builder/sort.go (hunk 5 above)
Comment (MarkRunWu) [reply pkg/server/skydb/pq/builder/sort.go]: It's not used, removed.

File: pkg/server/skydb/pq/builder/factory.go (hunk 4 above)
Comment (MarkRunWu) [reply pkg/server/skydb/pq/builder/factory.go]: Updated to not support expr.KeyPathComponents() > 2 for now same as the current predicate expression.

File: pkg/server/skydb/pq/builder/factory.go (hunk 3 above)
Comment (MarkRunWu) [reply pkg/server/skydb/pq/builder/factory.go]: Think the original factory owns joinedTables and extraFields and they are also referenced by assembling sort expression for sorting by reference.
So decide to extends the original factory capability and rename the `PredicateSqlizerFactory` to `SqlizerFactory`

File: pkg/server/skydb/pq/builder/factory.go (hunk 2 above)
Comment (MarkRunWu) [reply pkg/server/skydb/pq/builder/factory.go]: Create another method for sorting instead.

File: pkg/server/skydb/pq/builder/factory.go (hunk 1 above)
Comment (MarkRunWu) [reply pkg/server/skydb/pq/builder/factory.go]: Done at 31bf84d

**Comment by Steven-Chan** [comment]
@MarkRunWu Thanks for the update!

Though, imo, the `NewSort` function does not really fit into `SqlizerFactory` interface, but that's the problem of existing code (or maybe just me who thinks its a problem :scream:), which may need refactoring on the `SqlizerFactory`.

So I think this PR is good enough to merge.

**Pull request merged**